frontend/test/pytest/test_jax_dynamic_api.py

```diff
@@ -980,5 +1082,29 @@ def case():
     assert_array_and_dtype_equal(f(False, 3), jnp.zeros([4, 3]))
 
 
+def test_qjit_cond_capture():
+    """Test that catalyst tensor primitive is compatible with quantum conditional"""
+
+    @qjit
+    def f(flag, sz):
+        a = jnp.ones([sz, 3], dtype=float)
+
+        @cond(flag)
+        def case():
+            b = jnp.ones([sz, 3], dtype=float)
+            return a + b
+
+        @case.otherwise
+        def case():
+            b = jnp.ones([sz, 3], dtype=float)
+            return a + b
+
+        c = case()
```

Comment:
```
True (the default):  all dynamic dimension variables will be handled as Jaxpr constants. So

    The mentioned mixture would become possible.
    No dimension modification would be allowed inside loops

False would handle all the dimensions as variables instead.

    Mixing argument and captured arrays is not possible.
    Dimension modification within the loops are allowed

```

Hi Sergei, if I am reading this correctly, True means that a loop has been specialized to a particular context while False would allow us to reuse the loop for multiple variables of different lengths, is this correct? E.g.,

```
    @qjit()
    def f(sz):
        x = jnp.ones([sz], dtype=float)

        @while_loop(lambda i, _: i < 3)
        def loop(i, a):
            return i + 1, a + x

        _, a2 = loop(1, x)

        y = jnp.ones([sz + 1], dtype=float)
        _, a3 = loop(1, y)

        return a2
```

Is my interpretation of this correct?

Reply:
>  Is my interpretation of this correct?

Hmm, not quite. We repeat the tracing every time we call `loop(..)` in Python, so we do not specialize anything, I would say. It is just two independent loop instances, where the first one is correct, but the second one contains a shape mismatch error.

The switch chooses between two strategies of dealing with argument dimension variables.
Let me add some context. All the variables we see in the primitive body might be either arguments or constants. This is true for their dimensions as well, because they are variables too, just the implicit ones (which means - invisible in Python). Now, for dimensions of _constants_ there is no choice, they are just constants too. But what about the dimensions of _arguments_ ? For each of them we have two options:
* Treat the dimension of an argument as another _implicit argument_. This way we would allow writing mutable-dimension programs (with a few more precautions which are already present in the sources). But there is a downside - we would loose the ability to put an argument and a constant as arguments to a same-dimension operation like add, mul, etc. 
* Treat the dimension of an argument as another _constant_. This way we can mix arguments and constants in operations if they refer to the same dimensions. And the downside is - no more mutable dimension programs: all dimensions are constants.

We need to choose between the above options, we only see the whole picture when we finish the tracing. Since we strictly want a single-pass tracing algorithm, we need an additional information. This is what the flag does. It switches the strategy for all the dimensions of this primitive.

Reply:
(I have updated the PR description as well)

Reply:
> where the first one is correct, but the second one contains a shape mismatch error.

Will the second one contain a shape mismatch error or will it just run through the first three elements without generating a fourth?

Reply:
> Treat the dimension of an argument as another implicit argument. This way we would allow writing mutable-dimension programs (with a few more precautions which are already present in the sources). But there is a downside - we would loose the ability to mix arguments and constants in operations like add, mul, etc.

Mmm... yes I get this. I was wondering whether we only trace the loop once and reuse it (since under the hood it would be a function, right?). But it is retraced as you mentioned. Do we already have a test for this error?

Would it also be possible to have something like:

```
    @qjit()
    def f(sz):
        x = jnp.ones([sz], dtype=float)

        @while_loop(lambda i, _: i < x[i]) # <--- here is the dynamically shaped array.
        def loop(i, a):
            return i + 1, a + x

        _, a2 = loop(1, x)

        return a2
```

Reply:
> since under the hood it would be a function, right?

For jaxpr, all primitives are nameless functions. Tracing a loop results in emitting a primitive having its own body.  

> Would it also be possible to have something like

This example works fine

> Do we already have a test for this error?

Did you mention the error in the second loop of your original example? It is a shape mismatch error, we are trying to add `f64[sz]` with `f64[sz']` where `sz'` is `sz+1`